File: App/Resources/AlgorithmDetails/laziestsort/laziestsort.kt
fun binaryInsertionSort(arr: Array<Int>, lo: Int, hi: Int) {
  for (i in lo + 1 until hi) {
    val key = arr[i]
    var left = lo
    var right = i
    while (left < right) {
      val mid = (left + right) / 2
      if (arr[mid] <= key) {
        left = mid + 1
      } else {
        right = mid
      }
    }
    for (j in i downTo left + 1) {
      arr[j] = arr[j - 1]
    }
    arr[left] = key
  }
}

fun swapRange(arr: Array<Int>, a: Int, b: Int, length: Int) {
  for (i in 0 until length) {
    val t = arr[a + i]
    arr[a + i] = arr[b + i]
    arr[b + i] = t
  }
}

// Swaps the two adjacent blocks arr[lo, mid) and arr[mid, hi) so their order is
// reversed, using no auxiliary storage: the smaller of the two remaining pieces is
// always swapped whole against an equal-sized piece of the other, which shrinks one
// piece to nothing a little at a time until both are exhausted.
fun rotate(arr: Array<Int>, lo: Int, mid: Int, hi: Int) {
  var i = mid - lo
  var j = hi - mid
  if (i == 0 || j == 0) {
    return
  }
  while (i != j) {
    if (i < j) {
      swapRange(arr, mid - i, mid + j - i, i)
      j -= i
    } else {
      swapRange(arr, mid - i, mid, j)
      i -= j
    }
  }
  swapRange(arr, mid - i, mid, i)
}

// Finds the first index in [lo, hi) whose element is not less than value, by doubling
// the step size until it overshoots and then binary-searching the resulting bracket,
// rather than scanning one element at a time. Assumes arr[lo] < value.
fun gallop(arr: Array<Int>, lo: Int, hi: Int, value: Int): Int {
  var left = lo
  var step = 1
  var right = lo + step
  while (right < hi && arr[right] < value) {
    left = right
    step *= 2
    right = lo + step
  }
  right = minOf(right, hi)
  while (right - left > 1) {
    val mid = (left + right) / 2
    if (arr[mid] < value) {
      left = mid
    } else {
      right = mid
    }
  }
  return right
}

// Merges the sorted run arr[lo, mid) into the sorted run arr[mid, hi) in place. `left`
// tracks the first not-yet-placed element of the left run, and `right` tracks the start
// of the not-yet-consumed remainder of the right run.
fun merge(arr: Array<Int>, lo: Int, mid: Int, hi: Int) {
  var left = lo
  var right = mid
  while (left < right && right < hi) {
    if (arr[left] <= arr[right]) {
      left++
    } else {
      val boundary = gallop(arr, right, hi, arr[left])
      rotate(arr, left, right, boundary)
      left += boundary - right
      right = boundary
    }
  }
}

fun integerSqrt(n: Int): Int {
  var r = Math.sqrt(n.toDouble()).toInt()
  while ((r + 1).toLong() * (r + 1) <= n) {
    r++
  }
  while (r.toLong() * r > n) {
    r--
  }
  return r
}

fun sort(arr: Array<Int>) {
  val n = arr.size
  if (n <= 16) {
    binaryInsertionSort(arr, 0, n)
    return
  }

  val blockSize = maxOf(16, integerSqrt(n))
  var low = 0
  while (low < n) {
    binaryInsertionSort(arr, low, minOf(low + blockSize, n))
    low += blockSize
  }

  // Merge blocks back to front: the already-sorted run always starts at mergedStart,
  // and each step folds the block immediately before it into that run.
  val numBlocks = (n + blockSize - 1) / blockSize
  var mergedStart = (numBlocks - 1) * blockSize
  for (i in numBlocks - 2 downTo 0) {
    val leftStart = i * blockSize
    merge(arr, leftStart, mergedStart, n)
    mergedStart = leftStart
  }
}

fun main() {
  var array = arrayOf<Int>(
    55, 12, 84, 3, 47, 91, 26, 68, 8, 73, 40, 97, 15, 62, 34, 79, 21, 88, 5, 51, 66, 29, 44, 12,
  )
  sort(array)
  println("[%s]".format(array.joinToString(", ")))
}

File: Modules/BuiltInAlgorithms/Sources/Hybrid/LaziestSort.swift
import AlgorithmKit
import SortEngineKit

/// Ported from ArrayV's `LaziestSort` (`aphitorite`'s "Laziest Stable Sort") — a genuinely
/// in-place stable merge: `sqrt(n)`-sized blocks are binary-insertion-sorted first, then merged
/// one block at a time into the growing already-merged suffix via rotations instead of an
/// auxiliary buffer, using exponential (galloping) search to find each rotation's boundary in
/// `O(log(run length))` rather than a linear scan.
///
/// `rotate` is not re-derived here — it's exactly `GrailSortingTemplate.rotate(pos:lenA:lenB:)`'s
/// own multi-swap block rotation (`rotate(array, a, m, b)`'s `l`/`r`-shrinking loop is the
/// identical algorithm to `rotate(pos, lenA, lenB)`'s, just parameterized by endpoints instead of
/// lengths: `pos = a`, `lenA = m - a`, `lenB = b - m`), already fuzzed via 4 real algorithms in
/// this codebase's Grail cluster.
///
/// Two systematic translation decisions: `insertTo`/`binaryInsertion`'s comparisons are all
/// against a value captured once per outer iteration (bare, uncounted — same convention as every
/// other held-value binary search in this codebase, e.g. `MergeInsertionSort.blockSearch`);
/// `inPlaceMerge`'s own `array[i]` vs. `array[j]` comparison is between two live, independently-
/// marching positions, so it becomes a real `engine.compare`.
///
/// `stable: true` — matches ArrayV's own "Laziest **Stable** Sort" name, and both tie-breaking
/// rules confirm it: `binaryInsertion` uses `rightBinSearch` (`<`, never `<=`), which places a
/// new element after every existing equal one; `inPlaceMerge` only rotates elements out of the
/// left run when the left element is *strictly* greater than the right one, so a tie always just
/// advances `i`, leaving both elements exactly where a stable merge would. Every write is a mix
/// of `setValue` (`insertTo`'s shifts) and real swaps (`rotate`'s `multiSwap`), so — like the
/// other mixed-primitive hybrids in this codebase — the swap-tape-shadow-replay empirical check
/// doesn't apply; this is a by-construction argument instead.
///
/// Time/space bounds mirror the Grail cluster's own rotation-based in-place merges (same
/// amortized-rotation-cost argument, not independently re-derived): `O(n log n)` in every case,
/// `O(1)` space — genuinely in-place, no aux array anywhere.
public struct LaziestSort: SortAlgorithm {
  public let id = AlgorithmID(rawValue: "laziestsort")
  public let metadata = AlgorithmMetadata(
    displayName: "Laziest Stable",
    category: .hybrid,
    sizeRange: 16...256,
    growthModel: OperationGrowthModel(
      anchorSize: 1038, coefficients: [225199, 374.24, 0.128523],
      measuredSafeCeiling: nil),
    detectedGrowthModel: DetectedGrowthModel(
      family: .powerLog, coefficients: [0.552509, 1.58098], rSquared: 0.990328),
    stable: true,
    timeComplexity: ComplexityBounds(
      best: "O(n log n)", average: "O(n log n)", worst: "O(n log n)"),
    spaceComplexity: "O(1)",
    iconName: "moon.zzz.fill"
  )

  public init() {}

  public func record(into engine: inout RecordingEngine) {
    let n = engine.count
    guard n > 1 else { return }

    // Moves the element at `a` down to `b` (`b <= a`), shifting `[b, a)` right by one — ArrayV's
    // `insertTo`.
    func insertTo(_ a: Int, _ b: Int) {
      let temp = engine.values[a]
      var a = a
      while a > b {
        engine.setValue(a, engine.values[a - 1])
        a -= 1
      }
      engine.setValue(b, temp)
    }

    // Rightmost insertion point for `val` in the sorted range `[a, b)` — ties go right, which is
    // what keeps `binaryInsertion` stable.
    func rightBinSearch(_ aIn: Int, _ bIn: Int, _ val: Int) -> Int {
      var a = aIn
      var b = bIn
      while a < b {
        let mid = a + (b - a) / 2
        if val < engine.values[mid] {
          b = mid
        } else {
          a = mid + 1
        }
      }
      return a
    }

    // Leftmost insertion point for `val` in the sorted range `[a, b)`.
    func leftBinSearch(_ aIn: Int, _ bIn: Int, _ val: Int) -> Int {
      var a = aIn
      var b = bIn
      while a < b {
        let mid = a + (b - a) / 2
        if val <= engine.values[mid] {
          b = mid
        } else {
          a = mid + 1
        }
      }
      return a
    }

    // Exponential (galloping) search: doubles outward from `a` until it overshoots `val`'s
    // insertion point or runs off the end of `[a, b)`, then binary-searches within that bracket.
    // Finds the point in `O(log(distance))` instead of `leftBinSearch`'s `O(log(b - a))` when the
    // real answer is close to `a` — which it usually is here, since `inPlaceMerge` only ever
    // calls this to find where one element from the left run belongs among the right run.
    func leftExpSearch(_ a: Int, _ b: Int, _ val: Int) -> Int {
      var i = 1
      while a - 1 + i < b && val > engine.values[a - 1 + i] {
        i *= 2
      }
      return leftBinSearch(a + i / 2, min(b, a - 1 + i), val)
    }

    // Binary insertion sort over `[a, b)`.
    func binaryInsertion(_ a: Int, _ b: Int) {
      guard a + 1 < b else { return }
      for i in (a + 1)..<b {
        insertTo(i, rightBinSearch(a, i, engine.values[i]))
      }
    }

    // Stably merges the already-sorted `[a, m)` into the already-sorted `[m, b)`, in place: `i`
    // walks the left run, `j` the right. On a tie or `array[i] <= array[j]`, `i` just advances —
    // already in order. When `array[i]` is strictly greater, everything in the right run up to
    // (but not including) where `array[i]` belongs is strictly smaller than it, so rotating
    // `[i, j)` with `[j, k)` moves that whole already-scanned left-run block past them in one
    // shot, instead of inserting one element at a time.
    func inPlaceMerge(_ a: Int, _ m: Int, _ b: Int) {
      var i = a
      var j = m
      while i < j && j < b {
        if engine.compare(i, j, by: >) {
          let value = engine.values[i]
          let k = leftExpSearch(j + 1, b, value)
          GrailSortingTemplate.rotate(&engine, i, j - i, k - j)
          i += k - j
          j = k
        } else {
          i += 1
        }
      }
    }

    if n <= 16 {
      binaryInsertion(0, n)
      return
    }

    let blockLen = max(16, Int(Double(n).squareRoot()))
    var i = 0
    while i + 2 * blockLen < n {
      binaryInsertion(i, i + blockLen)
      i += blockLen
    }
    binaryInsertion(i, n)

    while i - blockLen >= 0 {
      inPlaceMerge(i - blockLen, i, n)
      i -= blockLen
    }
  }
}
